From c42a48e6f33a57673112f5d65d3f30f6cb35a6cc Mon Sep 17 00:00:00 2001
From: Atchaya R <quic_ratchaya@quicinc.com>
Date: Tue, 30 May 2023 15:44:07 +0530
Subject: [PATCH] include: Add support for prebuilt packages

    If the package Makefile contains the line
    PKG_TARGETS:=ipkg-prebuilt
    openwrt will try to use a prebuilt ipkg found in the
    prebuilt/$(BOARD) or prebuilt/$(SUBTARGET) if SUBTARGET is defined;
    if this doesn't exist it will fall back to compiling the ipkg
    unless PKG_FORCE_PREBUILT is set

Change-Id: I6b0033e721b79385241b36e5aa8f49b35d219484
Signed-off-by: Atchaya R <quic_ratchaya@quicinc.com>
---
 include/package-ipkg-prebuilt.mk | 69 ++++++++++++++++++++++++++++++++
 include/package.mk               |  3 +-
 rules.mk                         |  1 +
 3 files changed, 72 insertions(+), 1 deletion(-)
 create mode 100755 include/package-ipkg-prebuilt.mk

diff --git a/include/package-ipkg-prebuilt.mk b/include/package-ipkg-prebuilt.mk
new file mode 100755
index 000000000000..8954c5c97a17
--- /dev/null
+++ b/include/package-ipkg-prebuilt.mk
@@ -0,0 +1,69 @@
+#
+# Based on include/package-ipkg.mk
+#
+# Copyright (c) 2013 The Linux Foundation. All rights reserved.
+# Copyright (C) 2006,2007 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
+#
+
+include $(INCLUDE_DIR)/feeds.mk
+
+ifeq ($(DUMP),)
+  define BuildTarget/ipkg-prebuilt
+    ifeq ($(PKG_FORCE_PREBUILT)$(wildcard $(PREBUILT_DIR)/$(1)_$(VERSION)_$(PKGARCH).ipk),)
+      $(BuildTarget/ipkg)
+    else
+      PKG_$(1):=$(1)_$(VERSION)_$(PKGARCH).ipk
+      PDIR_$(1):=$(call FeedPackageDir,$(1))
+      PRE_$(1):=$(PREBUILT_DIR)/$$(PKG_$(1))
+      IPKG_$(1):=$$(PDIR_$(1))/$$(PKG_$(1))
+      INFO_$(1):=$(IPKG_STATE_DIR)/info/$(1).list
+
+      Build/InstallDev:=
+
+      ifeq ($(if $(VARIANT),$(BUILD_VARIANT)),$(VARIANT))
+        ifdef Package/$(1)/install
+          ifneq ($(CONFIG_PACKAGE_$(1))$(SDK)$(DEVELOPER),)
+	    IPKGS += $(1)
+
+            ifeq ($(CONFIG_PACKAGE_$(1)),y)
+              .PHONY: $(PKG_INSTALL_STAMP).$(1)
+              compile: $$(IPKG_$(1)) $(PKG_INSTALL_STAMP).$(1)
+              $(PKG_INSTALL_STAMP).$(1):
+			if [ -f $(PKG_INSTALL_STAMP).clean ]; then \
+				rm -f \
+					$(PKG_INSTALL_STAMP) \
+					$(PKG_INSTALL_STAMP).clean; \
+			fi; \
+			echo "$(1)" >> $(PKG_INSTALL_STAMP)
+
+              install: $$(INFO_$(1))
+            else
+              compile: $$(IPKG_$(1))
+            endif
+          else
+            compile: $(1)-disabled
+            $(1)-disabled:
+		@echo "WARNING: skipping $(1) -- package not selected"
+          endif
+        endif
+      endif
+
+      $$(IPKG_$(1)):
+	mkdir -p $$(PDIR_$(1))
+	$(CP) $$(PRE_$(1)) $$@
+
+      $$(INFO_$(1)): $$(IPKG_$(1))
+	@[ -d $(TARGET_DIR)/tmp ] || mkdir -p $(TARGET_DIR)/tmp
+	$(OPKG) install $$(IPKG_$(1))
+	$(if $(filter-out essential,$(PKG_FLAGS)),for flag in $(filter-out essential,$(PKG_FLAGS)); do $(OPKG) flag $$$$flag $(1); done,$(OPKG) flag ok $(1))
+
+      $(1)-clean:
+	rm -f $$(PDIR_$(1))/$(1)_*
+
+      clean: $(1)-clean
+    endif
+  endef
+endif
diff --git a/include/package.mk b/include/package.mk
index 1dc6d98a74a7..81ed9b7abe94 100644
--- a/include/package.mk
+++ b/include/package.mk
@@ -132,6 +132,7 @@ PKG_INSTALL_STAMP:=$(PKG_INFO_DIR)/$(PKG_DIR_NAME).$(if $(BUILD_VARIANT),$(BUILD
 include $(INCLUDE_DIR)/package-defaults.mk
 include $(INCLUDE_DIR)/package-dumpinfo.mk
 include $(INCLUDE_DIR)/package-ipkg.mk
+include $(INCLUDE_DIR)/package-ipkg-prebuilt.mk
 include $(INCLUDE_DIR)/package-bin.mk
 include $(INCLUDE_DIR)/autotools.mk
 
@@ -325,7 +326,7 @@ endif
     $(if $(CHECK),,$(Dumpinfo/Package)), \
     $(foreach target, \
       $(if $(Package/$(1)/targets),$(Package/$(1)/targets), \
-        $(if $(PKG_TARGETS),$(PKG_TARGETS), ipkg) \
+        $(if $(PKG_TARGETS),$(PKG_TARGETS), ipkg-prebuilt) \
       ), $(BuildTarget/$(target)) \
     ) \
   )
diff --git a/rules.mk b/rules.mk
index dcac49667bbf..55867f94bfb1 100644
--- a/rules.mk
+++ b/rules.mk
@@ -115,6 +115,7 @@ BIN_DIR:=$(OUTPUT_DIR)/targets/$(BOARD)/$(SUBTARGET)
 INCLUDE_DIR:=$(TOPDIR)/include
 SCRIPT_DIR:=$(TOPDIR)/scripts
 BUILD_DIR_BASE:=$(TOPDIR)/build_dir
+PREBUILT_DIR:=$(TOPDIR)/prebuilt/$(if $(SUBTARGET),$(SUBTARGET),$(BOARD))
 ifeq ($(CONFIG_EXTERNAL_TOOLCHAIN),)
   GCCV:=$(call qstrip,$(CONFIG_GCC_VERSION))
   LIBC:=$(call qstrip,$(CONFIG_LIBC))
-- 
2.34.1

